L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc

```diff
@@ -149,9 +186,163 @@ void L1NNTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
   desc.add<int>("nparticles", 10);
   desc.add<double>("conesize", 0.4);
   desc.add<double>("seedpt", 20);
+  desc.add<bool>("HW", true);
+  desc.add<bool>("emseed", true);
+  desc.add<bool>("debug", false);
   desc.add<edm::InputTag>("L1PFObjects", edm::InputTag("L1PFProducer", "l1pfCandidates"));
   descriptions.add("L1NNTauProducer", desc);
 }
+
+void L1NNTauProducer::makeTau_HW(const l1t::PFCandidate& seed,
+                                 l1t::PFCandidateCollection& parts,
+                                 std::unique_ptr<l1t::PFTauCollection>& iTaus) {
+  // Seed Cone Jet algorithm with ap_fixed types and hardware emulation
+  L1TauEmu::detaphi_t rCone2 =
+      L1TauEmu::detaphi_t(fTauSize_ * fTauSize_ * L1TauEmu::etaphi_base * L1TauEmu::etaphi_base);
+  unsigned lId = 0;
+  L1TauEmu::pt_t pt_tot = 0;
+  input2_t p1_tot = 0;
+  input2_t p1x_tot = 0;
+  input2_t p1y_tot = 0;
+  input2_t p1z_tot = 0;
+
+  float p_tot = 0;
+  float px_tot = 0;
+  float py_tot = 0;
+  float pz_tot = 0;
+
+  float eta_1 = seed.eta();
+  float phi_1 = seed.phi();
+  input_t e1ta_1 = seed.eta();
+  input_t p1hi_1 = seed.phi();
+  L1TauEmu::pt_t pt = 0;
+  L1TauEmu::z0_t z0 = 0;
+  L1TauEmu::dxy_t dxy = 0;
+  for (unsigned i0 = 0; i0 < parts.size(); i0++) {
+    pt_tot = pt_tot + L1TauEmu::pt_t(parts[i0].pt());
+    if (L1TauEmu::inCone(seed, (parts[i0]), rCone2)) {
+      if (parts[i0].id() == l1t::PFCandidate::Electron || parts[i0].id() == l1t::PFCandidate::ChargedHadron ||
+          parts[i0].id() == l1t::PFCandidate::Photon) {
+        lId++;
+        pt = pt + L1TauEmu::pt_t(parts[i0].pt());
+        float deta = parts[i0].eta() - eta_1;
+        float dphi = parts[i0].phi() - phi_1;
+        float dr2 = deta * deta + dphi * dphi;
+        pz_tot = pz_tot + (parts[i0].pt()) * (1 - dr2 * 0.5);
+        py_tot = py_tot + (parts[i0].pt()) * dphi;  //sin(dphi ));
+        px_tot = px_tot + (parts[i0].pt()) * deta;  //sin(deta ));
```

Comment:
```suggestion
```
also these do not seem to be used anywhere...
After a very quick glance this code still need quite some cleaning